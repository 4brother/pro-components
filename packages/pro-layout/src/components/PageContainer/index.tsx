import {
  computed,
  defineComponent,
  unref,
  toRefs,
  type VNode,
  type VNodeChild,
  type PropType,
  type ExtractPropTypes,
} from 'vue'
/* replace antd ts define */
import omit from 'omit.js'

import PageHeader, { pageHeaderProps } from 'ant-design-vue/es/page-header'
import { Tabs, Affix, Spin } from 'ant-design-vue'
import type { TabPaneProps } from './interfaces/TabPane'
import type { TabBarExtraContent /*, TabsProps */ } from './interfaces/Tabs'
import type { AffixProps } from './interfaces/Affix'
/* replace antd ts define end */
import { type RouteContextProps, useRouteContext } from '../../RouteContext'
import { getSlot, getSlotVNode } from '../../utils'
import 'ant-design-vue/es/affix/style'
import 'ant-design-vue/es/page-header/style'
import 'ant-design-vue/es/breadcrumb/style'
import 'ant-design-vue/es/tabs/style'
import 'ant-design-vue/es/spin/style'
import GridContent from '../GridContent'
import FooterToolbar from '../FooterToolbar'
import { withInstall } from 'ant-design-vue/es/_util/type'
import type { CustomRender } from '../../typings'
import type { DefaultPropRender, PageHeaderRender } from '../../RenderTypings'
import './index.less'

export interface Tab {
  key: string
  tab: string | VNode | JSX.Element
}

export const pageHeaderTabConfig = {
  /**
   * @name tabs 的列表
   */
  tabList: {
    type: [Object, Function, Array] as PropType<(Omit<TabPaneProps, 'id'> & { key?: string })[]>,
    default: () => undefined,
  },
  /**
   * @name 当前选中 tab 的 key
   */
  tabActiveKey: String, //PropTypes.string,
  /**
   * @name tab 上多余的区域
   */
  tabBarExtraContent: {
    type: [Object, Function] as PropType<TabBarExtraContent>,
    default: () => undefined,
  },
  /**
   * @name tabs 的其他配置
   */
  tabProps: {
    type: Object, //as PropType<TabsProps>,
    default: () => undefined,
  },
  /**
   * @name 固定 PageHeader 到页面顶部
   */
  fixedHeader: Boolean, //PropTypes.looseBool,
  // events
  onTabChange: Function, //PropTypes.func,
}
export type PageHeaderTabConfig = Partial<ExtractPropTypes<typeof pageHeaderTabConfig>>

export const pageContainerProps = {
  ...pageHeaderTabConfig,
  ...pageHeaderProps,
  prefixCls: {
    type: String,
    default: 'ant-pro',
  }, //PropTypes.string.def('ant-pro'),
  title: {
    type: [Object, String, Boolean, Function] as PropType<DefaultPropRender>,
    default: () => null,
  },
  subTitle: {
    type: [Object, String, Boolean, Function] as PropType<DefaultPropRender>,
    default: () => null,
  },
  content: {
    type: [Object, String, Boolean, Function] as PropType<DefaultPropRender>,
    default: () => null,
  },
  extraContent: {
    type: [Object, String, Boolean, Function] as PropType<DefaultPropRender>,
    default: () => null,
  },
  header: {
    type: [Object, String, Boolean, Function] as PropType<DefaultPropRender>,
    default: () => null,
  },
  footer: {
    type: [Object, String, Boolean, Function] as PropType<DefaultPropRender>,
    default: () => null,
  },
  pageHeaderRender: {
    type: [Object, Function, Boolean] as PropType<PageHeaderRender>,
  },
  affixProps: {
    type: [Object, Function] as PropType<AffixProps>,
  },
  ghost: {
    type: Boolean,
    default: () => undefined,
  }, //PropTypes.looseBool,
  loading: {
    type: Boolean,
    default: () => undefined,
  }, //PropTypes.looseBool,
}

export type PageContainerProps = Partial<ExtractPropTypes<typeof pageContainerProps>>

const renderFooter = (
  props: Omit<
    PageContainerProps & {
      prefixedClassName: string
    },
    'title'
  >
): VNodeChild | JSX.Element => {
  const { tabList, tabActiveKey, onTabChange, tabBarExtraContent, tabProps, prefixedClassName } = props
  if (tabList && tabList.length) {
    return (
      <Tabs
        class={`${prefixedClassName}-tabs`}
        activeKey={tabActiveKey}
        onChange={(key: string | number) => {
          if (onTabChange) {
            onTabChange(key)
          }
        }}
        tabBarExtraContent={tabBarExtraContent}
        {...tabProps}
      >
        {tabList.map((item) => (
          <Tabs.TabPane {...item} tab={item.tab} key={item.key} />
        ))}
      </Tabs>
    )
  }
  return null
}

const renderPageHeader = (
  content: CustomRender,
  extraContent: CustomRender,
  prefixedClassName: string
): VNodeChild | JSX.Element | null => {
  if (!content && !extraContent) {
    return null
  }
  return (
    <div class={`${prefixedClassName}-detail`}>
      <div class={`${prefixedClassName}-main`}>
        <div class={`${prefixedClassName}-row`}>
          {content && (
            <div class={`${prefixedClassName}-content`}>{(typeof content === 'function' && content()) || content}</div>
          )}
          {extraContent && (
            <div class={`${prefixedClassName}-extraContent`}>
              {(typeof extraContent === 'function' && extraContent()) || extraContent}
            </div>
          )}
        </div>
      </div>
    </div>
  )
}

const defaultPageHeaderRender = (
  props: PageContainerProps,
  value: Required<RouteContextProps> & { prefixedClassName: string }
) => {
  const { title, tabList, tabActiveKey, content, pageHeaderRender, header, extraContent, ...restProps } = omit(props, [
    'prefixCls',
  ])
  if (pageHeaderRender) {
    return pageHeaderRender({ ...props, ...value })
  }
  let pageHeaderTitle = title
  if (!title && title !== false) {
    pageHeaderTitle = value.title
  }
  const unrefBreadcrumb = unref(value.breadcrumb || {})
  const breadcrumb = restProps.breadcrumb || {
    ...unrefBreadcrumb,
    routes: unrefBreadcrumb.routes,
    itemRender: unrefBreadcrumb.itemRender,
  }
  // inject value
  return (
    <PageHeader
      {...restProps}
      title={pageHeaderTitle}
      breadcrumb={breadcrumb}
      footer={renderFooter({
        ...restProps,
        tabList,
        tabActiveKey,
        prefixedClassName: value.prefixedClassName,
      })}
    >
      {header || renderPageHeader(content, extraContent, value.prefixedClassName)}
    </PageHeader>
  )
}

const PageContainer = defineComponent({
  name: 'PageContainer',
  inheritAttrs: false,
  props: pageContainerProps,
  setup(props, { slots }) {
    const { loading, affixProps, ghost } = toRefs(props)
    const value = useRouteContext()
    const { getPrefixCls } = value
    const prefixCls = props.prefixCls || getPrefixCls()
    const prefixedClassName = computed(() => `${prefixCls}-page-container`)
    const classNames = computed(() => {
      return {
        [prefixedClassName.value]: true,
        [`${prefixCls}-page-container-ghost`]: ghost.value,
      }
    })

    const headerDom = computed(() => {
      const tags = getSlotVNode<DefaultPropRender>(slots, props, 'tags')
      const headerContent = getSlotVNode<DefaultPropRender>(slots, props, 'content')
      const extra = getSlotVNode<DefaultPropRender>(slots, props, 'extra')
      const extraContent = getSlotVNode<DefaultPropRender>(slots, props, 'extraContent')
      return (
        <div class={`${prefixedClassName.value}-warp`}>
          {defaultPageHeaderRender(
            {
              ...props,
              tags,
              content: headerContent,
              extra,
              extraContent,
              prefixCls: undefined,
            },
            {
              ...value,
              prefixedClassName: prefixedClassName.value,
            }
          )}
        </div>
      )
    })

    return () => {
      const footer = getSlot(slots, props, 'footer')

      return (
        <div class={classNames.value}>
          {value.fixedHeader ? (
            <Affix offsetTop={value.fixedHeader ? value.headerHeight : 0} {...affixProps.value}>
              {headerDom.value}
            </Affix>
          ) : (
            headerDom.value
          )}
          <GridContent>
            {loading.value ? (
              <Spin />
            ) : slots.default ? (
              <div>
                <div class={`${prefixedClassName.value}-children-content`}>{slots.default()}</div>
                {value.hasFooterToolbar && (
                  <div
                    style={{
                      height: 48,
                      marginTop: 24,
                    }}
                  />
                )}
              </div>
            ) : null}
          </GridContent>
          {value.hasFooterToolbar && <FooterToolbar>{footer}</FooterToolbar>}
        </div>
      )
    }
  },
})
// <WaterMark content="Pro Layout"></WaterMark>
export default withInstall(PageContainer)
