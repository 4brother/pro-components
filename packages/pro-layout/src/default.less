@import 'ant-design-vue/es/style/themes/variable.less';

@basicLayout-prefix-cls: ~'@{ant-prefix}-pro-basicLayout';
@pro-layout-header-height: 48px;
