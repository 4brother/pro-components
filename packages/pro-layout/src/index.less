@import './BasicLayout.less';
@import './components/PageContainer/index.less';
