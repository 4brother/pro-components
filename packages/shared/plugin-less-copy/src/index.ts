import { resolve } from 'path'
import { normalizePath, type Plugin } from 'vite'

const excludeRegx = /node_modules/

function createPlugin(): Plugin {
  const maps = new Map<string, string>()
  const srcDir = normalizePath(resolve('./src/'))

  return {
    name: 'vite-plugin-less-copy',
    enforce: 'pre',
    apply: 'build',
    transform(code: string, id: string) {
      if (!id.endsWith('.less') || excludeRegx.test(id)) {
        return
      }

      maps.set(id, code)
      return code
    },

    generateBundle() {
      maps.forEach((code, file) => {
        this.emitFile({
          type: 'asset',
          fileName: file.replace(srcDir, '').substring(1),
          source: code,
        })
      })
    },
  }
}

export default createPlugin
