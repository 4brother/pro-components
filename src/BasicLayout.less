@import "~ant-design-vue/es/style/themes/default";

@sider-menu-prefix-cls: ~'@{ant-prefix}-pro-sider-menu';
@nav-header-height: @layout-header-height;

.basicLayout {

  .ant-layout-sider-children {
    height: 100%;
  }

  .trigger {
    font-size: 18px;
    line-height: 64px;
    padding: 0 24px;
    cursor: pointer;
    transition: color 0.3s;

    &:hover {
      color: #1890ff;
    }
  }

  .color-picker {
    margin: 10px 0;
  }
}
