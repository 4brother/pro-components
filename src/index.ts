export * from './RouteContext';
export * from './typings';
export * from './utils/getMenuData';
export { createContext, useContext } from './hooks/context';
export type { ContextType, CreateContext } from './hooks/context';

export { default as FooterToolbar } from './FooterToolbar';
export { default as GlobalFooter } from './GlobalFooter';
export { default as GridContent } from './GridContent';
export { WrapContent } from './WrapContent';
// export {
//   default as ProProvider,
//   defaultProProviderProps,
//   useProProvider,
//   ProProviderData,
// } from './ProProvider';
export { default as PageContainer } from './PageContainer';

export { default as SiderMenuWrapper } from './SiderMenu';
export type { SiderMenuWrapperProps } from './SiderMenu';

export { default as BaseMenu, baseMenuProps } from './SiderMenu/BaseMenu';
export type { BaseMenuProps } from './SiderMenu/BaseMenu';
export type { MenuMode, OpenEventHandler, SelectInfo } from './SiderMenu/typings';

export { default as WaterMark } from './WaterMark';
export type { WaterMarkProps } from './WaterMark';

export { default } from './BasicLayout';
export type { BasicLayoutProps } from './BasicLayout';
