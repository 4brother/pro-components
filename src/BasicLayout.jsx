import './BasicLayout.less'

import { Layout } from 'ant-design-vue'
import { ContainerQuery } from 'vue-container-query'
import { SiderMenuWrapper, GlobalFooter } from './components'
import { getComponentFromProp, isFun } from './utils/util'
import { SiderMenuProps } from './components/SiderMenu/SiderMenu'
import HeaderView, { HeaderViewProps } from './Header'
import WrapContent from './WrapContent'
import ConfigProvider from './components/ConfigProvider'

export const BasicLayoutProps = {
  ...SiderMenuProps,
  ...HeaderViewProps,
  locale: {
    type: String,
    default: 'en-US'
  },
  breadcrumbRender: {
    type: Function,
    default: () => undefined
  },
  disableMobile: {
    type: Boolean,
    default: false
  },
  mediaQuery: {
    type: Object,
    default: () => {}
  },
  handleMediaQuery: {
    type: Function,
    default: () => undefined
  },
  footerRender: {
    type: Function,
    default: () => undefined
  }
}

const MediaQueryEnum = {
  'screen-xs': {
    maxWidth: 575
  },
  'screen-sm': {
    minWidth: 576,
    maxWidth: 767
  },
  'screen-md': {
    minWidth: 768,
    maxWidth: 991
  },
  'screen-lg': {
    minWidth: 992,
    maxWidth: 1199
  },
  'screen-xl': {
    minWidth: 1200,
    maxWidth: 1599
  },
  'screen-xxl': {
    minWidth: 1600
  }
}

const headerRender = (h, props) => {
  if (props.headerRender === false) {
    return null
  }
  return <HeaderView { ...{ props } } />
}

const defaultI18nRender = (key) => key

const BasicLayout = {
  name: 'BasicLayout',
  functional: true,
  props: BasicLayoutProps,
  render (h, content) {
    const { props, children } = content
    const {
      layout,
      // theme,
      isMobile,
      collapsed,
      mediaQuery,
      handleMediaQuery,
      handleCollapse,
      contentWidth,
      siderWidth,
      fixSiderbar,
      i18nRender = defaultI18nRender
    } = props

    const footerRender = getComponentFromProp(content, 'footerRender')
    const rightContentRender = getComponentFromProp(content, 'rightContentRender')
    const collapsedButtonRender = getComponentFromProp(content, 'collapsedButtonRender')
    const menuHeaderRender = getComponentFromProp(content, 'menuHeaderRender')
    const isTopMenu = layout === 'topmenu'

    const cdProps = {
      ...props,
      hasSiderMenu: !isTopMenu,
      footerRender,
      menuHeaderRender,
      rightContentRender,
      collapsedButtonRender
    }

    return (
      <ConfigProvider i18nRender={i18nRender} contentWidth={contentWidth}>
        <ContainerQuery query={MediaQueryEnum} onChange={handleMediaQuery}>
          <Layout class={{
            'ant-pro-basicLayout': true,
            'ant-pro-topmenu': isTopMenu,
            ...mediaQuery
          }}>
            <SiderMenuWrapper
              { ...{ props: cdProps } }
              collapsed={collapsed}
              onCollapse={handleCollapse}
            />
            <Layout class={[layout]} style={{
              paddingLeft:  fixSiderbar ? `${siderWidth}px` : '0',
              minHeight: '100vh'
            }}>
              {headerRender(h, {
                ...cdProps,
                mode: 'horizontal',
              })}
              <WrapContent class="ant-pro-basicLayout-content" contentWidth={contentWidth}>
                {children}
              </WrapContent>
              <Layout.Footer>
                { footerRender && (
                  isFun(footerRender) && footerRender(h) || footerRender
                ) || (
                  <GlobalFooter>
                    <template slot="links">
                      <a href="https://www.github.com/vueComponent/" target="_self">Github</a>
                      <a href="https://www.github.com/sendya/" target="_self">@Sendya</a>
                    </template>
                    <template slot="copyright">
                      <a href="https://github.com/vueComponent">vueComponent</a>
                    </template>
                  </GlobalFooter>
                )}
              </Layout.Footer>
            </Layout>
          </Layout>
        </ContainerQuery>
      </ConfigProvider>
    )
  }
}

export default BasicLayout
