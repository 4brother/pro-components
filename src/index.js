import BasicLayout, { BasicLayoutProps } from './BasicLayout'
import BlockLayout from './BlockLayout'
import PageView from './PageView'
import DocumentTitle from './components/DocumentTitle'
import { updateTheme } from './utils/dynamicTheme'

export {
  PageView,
  BlockLayout,
  DocumentTitle,
  BasicLayoutProps,

  updateTheme
}

export default BasicLayout
