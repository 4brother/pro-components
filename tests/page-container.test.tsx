import { mount } from '@vue/test-utils';
import { PageContainer } from '../src/PageContainer';
import { Tag, Button } from 'ant-design-vue';

describe('PageContainer', () => {

  const routes = [
    {
      path: 'index',
      breadcrumbName: 'First-level Menu',
    },
    {
      path: 'first',
      breadcrumbName: 'Second-level Menu',
    },
    {
      path: 'second',
      breadcrumbName: 'Third-level Menu',
    },
  ];
  const props = {
    title: 'Title',
    subTitle: "This is a subtitle",
    breadcrumb: { routes },
    onBack: () => {},
    tags: ['Tag 1', 'Tag 2'].map(tag => (<Tag color="blue">{tag}</Tag>)),
    extra: [<Button key="1" type="primary">主操作</Button>,],
    content: (<div>content</div>),
    extraContent: (<div>extraContent</div>),
    footer: [
      <Button key="3">重置</Button>,
      <Button key="2" type="primary">
        提交
      </Button>,
    ],
  }

  it('🥩 base use', () => {
    const wrapper = mount({
      render() {
        return (
          <PageContainer {...props}>
            <div>Page Content</div>
          </PageContainer>
        );
      },
    });
    console.log(wrapper.html());
  });

});
