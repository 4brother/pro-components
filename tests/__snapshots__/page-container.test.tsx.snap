// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageContainer ü•© base use 1`] = `
<div class="_rawValue __v_isRef _value">
  <div class="ant-pro-page-container-warp">
    <div class="ant-page-header has-breadcrumb ant-page-header-ghost">
      <div class="ant-breadcrumb"><span><span class="ant-breadcrumb-link"><a href="#/index">First-level Menu</a></span><span class="ant-breadcrumb-separator">/</span></span><span><span class="ant-breadcrumb-link"><a href="#/index/first">Second-level Menu</a></span><span class="ant-breadcrumb-separator">/</span></span><span><span class="ant-breadcrumb-link"><span>Third-level Menu</span></span><span class="ant-breadcrumb-separator">/</span></span></div>
      <div class="ant-page-header-heading">
        <div class="ant-page-header-back">
          <div role="button" tabindex="0" class="ant-page-header-back-button" aria-label="Back" style="border: 0px; background: transparent; padding: 0px; line-height: inherit; display: inline-block;"><span role="img" aria-label="arrow-left" class="anticon anticon-arrow-left"><svg class="" data-icon="arrow-left" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896" focusable="false"><path d="M872 474H286.9l350.2-304c5.6-4.9 2.2-14-5.2-14h-88.5c-3.9 0-7.6 1.4-10.5 3.9L155 487.8a31.96 31.96 0 000 48.3L535.1 866c1.5 1.3 3.3 2 5.2 2h91.5c7.4 0 10.8-9.2 5.2-14L286.9 550H872c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"></path></svg></span></div>
        </div>
        <!----><span class="ant-page-header-heading-title">Title</span><span class="ant-page-header-heading-sub-title">This is a subtitle</span><span class="ant-page-header-heading-tags"><span class="ant-tag ant-tag-blue">Tag 1<!----></span><span class="ant-tag ant-tag-blue">Tag 2<!----></span></span><span class="ant-page-header-heading-extra"><button class="ant-btn ant-btn-primary" type="button"><!----><span>‰∏ªÊìç‰Ωú</span></button></span>
      </div>
      <div class="ant-page-header-content">
        <div class="ant-pro-page-container-detail">
          <div class="ant-pro-page-container-main">
            <div class="ant-pro-page-container-row">
              <div class="ant-pro-page-container-content">
                <div>content</div>
              </div>
              <div class="ant-pro-page-container-extraContent">
                <div>extraContent</div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <!---->
    </div>
  </div>
  <div class="ant-pro-grid-content">
    <div class="ant-pro-grid-content-children">
      <div>
        <div class="ant-pro-page-container-children-content">
          <div>PageContent</div>
        </div>
        <!---->
      </div>
    </div>
  </div>
  <div class="ant-pro-footer-bar">
    <div class="ant-pro-footer-bar-left">
      <!---->
    </div>
    <div class="ant-pro-footer-bar-right"><button class="ant-btn" type="button">
        <!----><span>Èáç ÁΩÆ</span>
      </button><button class="ant-btn ant-btn-primary" type="button">
        <!----><span>Êèê ‰∫§</span>
      </button></div>
  </div>
</div>
`;
